program-runtime/src/executor_cache.rs

```diff
@@ -27,81 +27,95 @@ pub trait Executor: Debug + Send + Sync {
     ) -> Result<(), InstructionError>;
 }
 
-pub type Executors = HashMap<Pubkey, TransactionExecutor>;
-
+/// Relation between a LocalExecutorCacheEntry and its matching GlobalExecutorCacheEntry
 #[repr(u8)]
-#[derive(PartialEq, Debug)]
-enum TransactionExecutorStatus {
-    /// Executor was already in the cache, no update needed
-    Cached,
-    /// Executor was missing from the cache, but not updated
-    Missing,
-    /// Executor is for an updated program
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum LocalGlobalDifference {
+    /// The LocalExecutorCacheEntry did not change and is the same as the GlobalExecutorCacheEntry.
+    None,
+    /// The LocalExecutorCacheEntry was inserted, no matching GlobalExecutorCacheEntry exists, so it needs to be inserted.
+    Inserted,
+    /// The LocalExecutorCacheEntry was replaced, the matching GlobalExecutorCacheEntry needs to be updated.
     Updated,
 }
 
-/// Tracks whether a given executor is "dirty" and needs to updated in the
-/// executors cache
+/// An entry of the LocalExecutorCache
 #[derive(Debug)]
-pub struct TransactionExecutor {
-    pub(crate) executor: Arc<dyn Executor>,
-    status: TransactionExecutorStatus,
+pub struct LocalExecutorCacheEntry {
+    executor: Arc<dyn Executor>,
+    difference: LocalGlobalDifference,
 }
 
-impl TransactionExecutor {
-    /// Wraps an executor and tracks that it doesn't need to be updated in the
-    /// executors cache.
-    pub fn new_cached(executor: Arc<dyn Executor>) -> Self {
-        Self {
-            executor,
-            status: TransactionExecutorStatus::Cached,
-        }
-    }
-
-    /// Wraps an executor and tracks that it needs to be updated in the
-    /// executors cache.
-    pub fn new_miss(executor: Arc<dyn Executor>) -> Self {
-        Self {
-            executor,
-            status: TransactionExecutorStatus::Missing,
-        }
-    }
+/// A subset of the GlobalExecutorCache containing only the executors relevant to one transaction
+#[derive(Default, Debug)]
+pub struct LocalExecutorCache {
+    pub executors: HashMap<Pubkey, LocalExecutorCacheEntry>,
+}
 
-    /// Wraps an executor and tracks that it needs to be updated in the
-    /// executors cache only if the transaction succeeded.
-    pub fn new_updated(executor: Arc<dyn Executor>) -> Self {
+impl LocalExecutorCache {
+    pub fn new(executable_keys: impl Iterator<Item = (Pubkey, Arc<dyn Executor>)>) -> Self {
```

Comment:
I think here and in `put` the API would look nicer if it took `IntoIterator` https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html. Then it would work for both iterators and collections